### PR TITLE
Implement respawn anchor, block explosions, fix bed explosion, add new events, improves PlayerRespawnEvent

### PR DESCRIPTION
**Pull request by joserobjr**
Add new events.

## TODO

- [x] Tests in game

### CODE REVIEW

File: src/main/java/cn/nukkit/Player.java [hunk 1]
```diff
@@ -4276,6 +4325,21 @@ public void kill() {
             this.dataPacket(pk);
         }
     }
+    
+    @PowerNukkitOnly
+    @Since("1.4.0.0-PN")
+    protected boolean isValidRespawnBlock(Block block) {
+        if (block.getId() == BlockID.RESPAWN_ANCHOR && block.getLevel().getDimension() == Level.DIMENSION_NETHER) {
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/Player.java:4332]: In that if, I think you forgot to add The End as valid for the Respawn Anchor.

File: src/main/java/cn/nukkit/block/BlockBed.java [hunk 2]
```diff
@@ -124,39 +122,63 @@ public boolean onActivate(@Nonnull Item item) {
...
 
-        if (!(this.level.getDimension() == Level.DIMENSION_OVERWORLD)) {
-            CompoundTag tag = EntityPrimedTNT.getDefaultNBT(this).putShort("Fuse", 0);
-            new EntityPrimedTNT(this.level.getChunk(this.getFloorX() >> 4, this.getFloorZ() >> 4), tag);
-            return true;
-        }
-
         BlockFace dir = getBlockFace();
+
+        boolean shouldExplode = this.level.getDimension() == Level.DIMENSION_OVERWORLD;
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/block/BlockBed.java:127]: You accidentally mistook the Respawn Anchor for the bed. The bed is to explode in other dimensions and the respawn anchor in the overworld.

File: src/main/java/cn/nukkit/block/BlockRespawnAnchor.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            return false;
+        }
+        
+        return attemptToSetSpawn(player);
+    }
+
+    @Since("1.4.0.0-PN")
+    @PowerNukkitOnly
+    protected boolean attemptToSetSpawn(@Nonnull Player player) {
+        if (this.level.getDimension() != Level.DIMENSION_OVERWORLD) {
```
Comment (LoboMetalurgico) [inline src/main/java/cn/nukkit/block/BlockRespawnAnchor.java:97]: You accidentally mistook the Respawn Anchor for the bed. The bed is to explode in other dimensions and the respawn anchor in the overworld.

**Review by LoboMetalurgico — CHANGES_REQUESTED** [review]
Thank you for your pull request! This PR is very good, but I found some problems that need to be corrected before we can merge.

File: src/main/java/cn/nukkit/Player.java (hunk 1 above)
Comment (RORIdev) [reply src/main/java/cn/nukkit/Player.java:4334]: This assumption is not correct. Accourding to the [wiki](https://minecraft.gamepedia.com/Respawn_Anchor#Exploding) : 
`If the player attempts to set their spawn at a charged respawn anchor in the Overworld, the End, or customized dimensions in which they are disabled, the block explodes (and is destroyed) similarly to when a bed is used in the Nether or the End. The explosion has a power of 5, and sets fire to surrounding blocks.`

File: src/main/java/cn/nukkit/block/BlockBed.java (hunk 2 above)
Comment (joserobjr) [reply src/main/java/cn/nukkit/block/BlockBed.java:129]: Thank you, imagine if it got into production, the reaction of the people after trying to sleep at night haha

**Review by LoboMetalurgico — APPROVED** [review]
Thank You!

**Pull request merged**